### PR TITLE
[NOD-828] Reimplement FFLDB

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: database2/ffldb/ff/rollback.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	// If the rollback point is greater than the current write cursor then
+	// something has gone very wrong, e.g. database corruption.
+	if cursor.currentFileNumber < targetFileNumber ||
+		(cursor.currentFileNumber == targetFileNumber && cursor.currentOffset < targetFileOffset) {
+		return errors.Errorf("targetLocation is greater than the " +
+			"current write cursor")
+	}
+
+	// Regardless of any failures that happen below, reposition the write
+	// cursor to the old flat file and offset.
```
Comment (svarogg) [inline database2/ffldb/ff/rollback.go]: old -> target

File: database2/ffldb/ff/rollback.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		openFile, err := s.openWriteFile(cursor.currentFileNumber)
+		if err != nil {
+			cursor.currentFile.Unlock()
+			log.Warnf("ROLLBACK: %s", err)
+			return nil
+		}
+		cursor.currentFile.file = openFile
+	}
+
+	// Truncate the to the provided rollback offset.
```
Comment (svarogg) [inline database2/ffldb/ff/rollback.go]: Truncate the what?

File: database2/ffldb/ff/write.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		// with LRU tracking. The close is done under the write lock
+		// for the file to prevent it from being closed out from under
+		// any readers currently reading from it.
+		cursor.Lock()
+		cursor.currentFile.Lock()
+		if cursor.currentFile.file != nil {
+			_ = cursor.currentFile.file.Close()
+			cursor.currentFile.file = nil
+		}
+		cursor.currentFile.Unlock()
```
Comment (svarogg) [inline database2/ffldb/ff/write.go]: I see that we have a ton of Unlocks without defer, so this is last time I write about it, simply go over the code and make sure that unlocks are always deferred.

File: database2/ffldb/ldb/leveldb.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	ldb, err := leveldb.OpenFile(path, nil)
+
+	// If the database is corrupted, attempt to recover.
+	if _, corrupted := err.(*errors.ErrCorrupted); corrupted {
+		log.Warnf("LevelDB corruption detected for path %s: %s",
+			path, err)
+		var err error
+		ldb, err = leveldb.RecoverFile(path, nil)
+		if err != nil {
+			return nil, err
```
Comment (svarogg) [inline database2/ffldb/ldb/leveldb.go]: Please wrap any error coming out of the system with our own message.

File: database2/ffldb/ldb/transaction.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
+package ldb
+
+import (
+	"errors"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// LevelDBTransaction is a thin wrapper around native leveldb
+// batches and snapshots. It supports both get and put.
+// Note: transactions provide data consistency over the state of
```
Comment (svarogg) [inline database2/ffldb/ldb/transaction.go]: Add empty line before the Note.
Also: typo: Note: **T**ransactions

File: database2/ffldb/ldb/transaction.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		return nil, err
+	}
+	batch := new(leveldb.Batch)
+
+	transaction := &LevelDBTransaction{
+		ldb:      db.ldb,
+		snapshot: snapshot,
+		batch:    batch,
+
```
Comment (svarogg) [inline database2/ffldb/ldb/transaction.go]: redundant empty line

File: database2/ffldb/ldb/transaction.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+// Commit commits whatever changes were made to the database
+// within this transaction.
+func (tx *LevelDBTransaction) Commit() error {
+	if tx.isClosed {
+		return errors.New("cannot commit a closed transaction")
+	}
+
+	tx.isClosed = true
+	tx.snapshot.Release()
```
Comment (svarogg) [inline database2/ffldb/ldb/transaction.go]: Can you explain (or provide a link) how does the relationship between snapshot and batch works

File: database2/transaction.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package database2
+
+// Transaction defines the interface of a generic kaspad database
+// transaction.
+// Note: transactions provide data consistency over the state of
+// the database as it was when the transaction started. There is
+// NO guarantee that if one puts data into the transaction then
+// it will be available to get within the same transaction.
+type Transaction interface {
```
Comment (svarogg) [inline database2/transaction.go]: I'd also love to have RollbackUnlessCommited if possible

File: dbaccess/block.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	// if there's an error.
+	bytes, err := block.Bytes()
+	if err != nil {
+		return err
+	}
+	blockLocation, err := db.AppendToStore(blockStoreName, bytes)
+	if err != nil {
+		rollbackErr := rollback()
+		if rollbackErr != nil {
+			return rollbackErr
```
Comment (svarogg) [inline dbaccess/block.go]: The original error is more important than the rollback error.
Try to somehow combine if possible (Shouldn't be too hard I think), but if not - return the original error

File: dbaccess/block.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	// Write the block's hash to the blockLocations bucket and
+	// rollback if there's an error.
+	blockLocationsKey := blockLocationKey(hash)
+	err = db.Put(blockLocationsKey, blockLocation)
+	if err != nil {
+		rollbackErr := rollback()
+		if rollbackErr != nil {
+			return rollbackErr
```
Comment (svarogg) [inline dbaccess/block.go]: Here too

File: database2/database.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
+package database2
+
+// Database defines the interface of a generic kaspad database.
+type Database interface {
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/database.go]: Created a ticket.

File: database2/ffldb/ldb/leveldb_test.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	// Rollback the transaction
+	err = tx.Rollback()
+	if err != nil {
+		t.Fatalf("TestLevelDBTransactionSanity: rollback "+
+			"returned unexpected error: %s", err)
+	}
+
+	// Make sure that the put data and the get data are equal
```
Comment (someone235) [inline database2/ffldb/ldb/leveldb_test.go]: I think it'll be clearer if you will check it before the rollback

File: database2/ffldb/ldb/transaction.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"errors"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// LevelDBTransaction is a thin wrapper around native leveldb
+// batches and snapshots. It supports both get and put.
+// Note: transactions provide data consistency over the state of
+// the database as it was when the transaction started. There is
+// NO guarantee that if one puts data into the transaction then
```
Comment (someone235) [inline database2/ffldb/ldb/transaction.go]: There's actually guarantee that it won't happen, no?

Comment (someone235) [reply database2/ffldb/ldb/transaction.go]: And btw, did you check that it doesn't happen in our codebase?

File: dbaccess/block.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+		return err
+	}
+	if exists {
+		return errors.Errorf("block %s already exists", hash)
+	}
+
+	// Save a reference to the current block location in case
+	// we fail and need to rollback.
+	previousBlockLocation := db.CurrentStoreLocation(blockStoreName)
+	rollback := func() error {
```
Comment (someone235) [inline dbaccess/block.go]: Why it's not part of transaction rollback?

File: database2/database.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	// Put sets the value for the given key. It overwrites
+	// any previous value for that key.
+	Put(key []byte, value []byte) error
+
+	// Get gets the value for the given key. It returns an
+	// error if the given key does not exist.
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/database.go]: Made it return a nil.

File: database2/ffldb/ffldb.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.updateCurrentStoreLocation(storeName, location)
+	if err != nil {
+		return nil, err
+	}
+
+	return location, err
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ffldb.go]: You're absolutely right. Fixed.

File: database2/ffldb/ff/write.go (hunk 3 above)
Comment (stasatdaglabs) [reply database2/ffldb/ff/write.go]: Created a ticket.

File: database2/ffldb/ldb/leveldb.go (hunk 4 above)
Comment (stasatdaglabs) [reply database2/ffldb/ldb/leveldb.go]: Created a ticket.

File: database2/ffldb/ff/read.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+	// When the requested flat file is open for writes, return it.
+	cursor := s.writeCursor
+	cursor.RLock()
+	if fileNumber == cursor.currentFileNumber && cursor.currentFile.file != nil {
+		openFile := cursor.currentFile
+		openFile.RLock()
+		cursor.RUnlock()
+		return openFile, nil
+	}
+	cursor.RUnlock()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/read.go]: Created a ticket.

File: database2/ffldb/ff/read.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+		// out from under them.
+		oldFile.Lock()
+		_ = oldFile.file.Close()
+		oldFile.Unlock()
+
+		delete(s.openFiles, lruFileNumber)
+		delete(s.fileNumberToLRUElement, lruFileNumber)
+	}
+	s.fileNumberToLRUElement[fileNumber] = lruList.PushFront(fileNumber)
+	s.lruMutex.Unlock()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/read.go]: Created a ticket.

File: database2/bucket.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
+package database2
+
+var separator = []byte("/")
+
+// Bucket is a helper struct meant to combine buckets,
+// sub-buckets, and keys into a single full key-value
+// database key.
+type Bucket struct {
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/bucket.go]: We spoke about this, and decided to leave it as is.

File: database2/ffldb/ff/flatfile.go [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+		storeName, fileNumber, fileLength)
+	return fileNumber, fileLength
+}
+
+// flatFilePath return the file path for the provided store's flat file number.
+func flatFilePath(dbPath string, storeName string, fileNumber uint32) string {
+	// Choose 9 digits of precision for the filenames. 9 digits provide
+	// 10^9 files @ 512MiB each a total of ~476.84PiB.
+
+	fileName := fmt.Sprintf("%s-%09d.fdb", storeName, fileNumber)
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/flatfile.go]: We spoke about this and decided to keep it as is.

File: database2/ffldb/ff/rollback.go [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+	}()
+
+	log.Debugf("ROLLBACK: Rolling back to file %d, offset %d",
+		targetFileNumber, targetFileOffset)
+
+	// Close the current write file if it needs to be deleted. Then delete
+	// all files that are newer than the provided rollback file while
+	// also moving the write cursor file backwards accordingly.
+	if cursor.currentFileNumber > targetFileNumber {
+		cursor.currentFile.Lock()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/rollback.go]: Created a ticket for defers.

File: database2/ffldb/ff/read.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+		return openFile, nil
+	}
+	cursor.RUnlock()
+
+	// Try to return an open file under the overall files read lock.
+	s.openFilesMutex.RLock()
+	if openFile, ok := s.openFiles[fileNumber]; ok {
+		s.lruMutex.Lock()
+		s.openFilesLRU.MoveToFront(s.fileNumberToLRUElement[fileNumber])
+		s.lruMutex.Unlock()
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/read.go]: Created a ticket for defers.

File: database2/ffldb/ldb/transaction.go (hunk 13 above)
Comment (stasatdaglabs) [reply database2/ffldb/ldb/transaction.go]: > And btw, did you check that it doesn't happen in our codebase?

Yep. Never happens as far as I can tell.

File: database2/ffldb/ff/read.go [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+	return data[4 : n-4], nil
+}
+
+// flatFile attempts to return an existing file handle for the passed flat file
+// number if it is already open as well as marking it as most recently used. It
+// will also open the file when it's not already open subject to the rules
+// described in openFile.
+//
+// NOTE: The returned flat file will already have the read lock acquired and
+// the caller MUST call .RUnlock() to release it once it has finished all read
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/read.go]: > I think it's better to return something like `tearDown` function to make the necessity of RUnlocking clearer, and also encapsulate this behaviour by making an additional function `func (s *flatFileStore) readFromFlatFileWithoutCheckSum(location *flatFileLocation)` (You probably don't want to use this name)

I've created a ticket to handle unlocks. I've added the above comment to the ticket to not forget about it.

**Comment by stasatdaglabs** [comment]
Rebuild after update in master

File: database2/ffldb/ff/read.go [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+// The function also keeps track of the open files, performs least recently
+// used tracking, and limits the number of open files to maxOpenFiles by closing
+// the least recently used file as needed.
+//
+// This function MUST be called with the open files mutex (s.openFilesMutex)
+// locked for WRITES.
+func (s *flatFileStore) openFile(fileNumber uint32) (*lockableFile, error) {
+	// Open the appropriate file as read-only.
+	filePath := flatFilePath(s.basePath, s.storeName, fileNumber)
+	file, err := os.Open(filePath)
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply database2/ffldb/ff/read.go]: The openFilesMutex mutex is supposed to protect against that.
The comment on this function says that it must be called with openFilesMutex locked, so I think it should be fine as it is right now.

File: dbaccess/block.go [hunk 25]
```diff
@@ -32,37 +32,20 @@ func StoreBlock(context Context, block *util.Block) error {
...
-	// Write the block's bytes to the block store and rollback
-	// if there's an error.
+	// Write the block's bytes to the block store
 	bytes, err := block.Bytes()
 	if err != nil {
 		return err
 	}
 	blockLocation, err := db.AppendToStore(blockStoreName, bytes)
 	if err != nil {
-		rollbackErr := rollback()
```
Comment (svarogg) [inline dbaccess/block.go]: You still want to rollback, no?

File: database2/bucket.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (svarogg) [reply database2/bucket.go]: There's a good reason to do this struct and not [][]byte: this way the path parameter is unexported.

File: database2/ffldb/ff/write.go [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+		cursor.currentFile.file = file
+	}
+
+	originalOffset := cursor.currentOffset
+	hasher := crc32.New(castagnoli)
+	var scratch [4]byte
+
+	// Data length.
+	byteOrder.PutUint32(scratch[:], dataLength)
+	err := s.writeData(scratch[:], "data length")
```
[the remark this answers is not in the document]
Comment (svarogg) [reply database2/ffldb/ff/write.go]: Correct me if I'm wrong: the location is only written in levelDB, it's nowhere to be found in the DB file.

Therefore I'd say it's a good idea to still leave this info inside the flat-file db. Then if database is corrupted, yet someone still wants to parse the flat-file for whatever reason (such as debugging) - he will be able to do so.

File: dbaccess/block.go (hunk 25 above)
Comment (stasatdaglabs) [reply dbaccess/block.go]: Rollbacks are now generic over any data store, and such are implemented inside ffldb.

File: dbaccess/block.go [hunk 27]
```diff
@@ -96,6 +79,9 @@ func FetchBlock(context Context, hash *daghash.Hash) (*util.Block, error) {
 	if err != nil {
 		return nil, err
 	}
+	if blockLocation == nil {
```
Comment (someone235) [inline dbaccess/block.go]: Isn't it weird that in one place we return nil if not found, and in the other we return error?

File: dbaccess/context.go [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"github.com/kaspanet/kaspad/database2"
+)
+
+// Context is an interface type representing the context in which queries run, currently relating to the
+// existence or non-existence of a database transaction
+// Call `.NoTx()` or `.NewTx()` to acquire a Context
+type Context interface {
+	db() (database2.DataAccessor, error)
```
Comment (someone235) [inline dbaccess/context.go]: Can you change to accessor()?

**Comment by stasatdaglabs** [comment]
Rebuild after update in master

File: database2/dataaccessor.go [hunk 29]
```diff
@@ -7,9 +7,9 @@ type DataAccessor interface {
 	// any previous value for that key.
 	Put(key []byte, value []byte) error
 
-	// Get gets the value for the given key. It returns nil if
+	// Get gets the value for the given key. It returns false if
 	// the given key does not exist.
-	Get(key []byte) ([]byte, error)
+	Get(key []byte) ([]byte, bool, error)
```
Comment (svarogg) [inline database2/dataaccessor.go]: Give names to the return values, to make it clear what each one means.

File: database2/ffldb/ffldb.go [hunk 30]
```diff
@@ -63,10 +63,10 @@ func (db *ffldb) Put(key []byte, value []byte) error {
 	return db.ldb.Put(key, value)
 }
 
-// Get gets the value for the given key. It returns nil if
+// Get gets the value for the given key. It returns false if
 // the given key does not exist.
 // This method is part of the DataAccessor interface.
-func (db *ffldb) Get(key []byte) ([]byte, error) {
+func (db *ffldb) Get(key []byte) ([]byte, bool, error) {
```
Comment (svarogg) [inline database2/ffldb/ffldb.go]: Here too

File: database2/ffldb/ldb/leveldb.go [hunk 31]
```diff
@@ -52,17 +52,17 @@ func (db *LevelDB) Put(key []byte, value []byte) error {
 	return errors.WithStack(db.ldb.Put(key, value, nil))
 }
 
-// Get gets the value for the given key. It returns nil if
+// Get gets the value for the given key. It returns false if
 // the given key does not exist.
-func (db *LevelDB) Get(key []byte) ([]byte, error) {
+func (db *LevelDB) Get(key []byte) ([]byte, bool, error) {
```
Comment (svarogg) [inline database2/ffldb/ldb/leveldb.go]: And here

File: dbaccess/block.go [hunk 32]
```diff
@@ -55,39 +55,43 @@ func StoreBlock(context Context, block *util.Block) error {
...
-	return db.Has(blockLocationsKey)
+	return accessor.Has(blockLocationsKey)
 }
 
 // FetchBlock returns the block of the given hash. Returns an
 // error if the block had not been previously inserted into the
 // database.
-func FetchBlock(context Context, hash *daghash.Hash) (*util.Block, error) {
-	db, err := context.db()
+func FetchBlock(context Context, hash *daghash.Hash) (*util.Block, bool, error) {
```
Comment (svarogg) [inline dbaccess/block.go]: And here, and anywhere else that I forgot to comment

**Comment by stasatdaglabs** [comment]
Rebuild after update in master

**Pull request merged**